docs/ref/models/indexes.txt

```diff
@@ -92,3 +92,44 @@ opclasses=['jsonb_path_ops'])`` creates a gin index on ``jsonfield`` using
 ``opclasses`` are ignored for databases besides PostgreSQL.
 
 :attr:`Index.name` is required when using ``opclasses``.
+
+``condition``
+-------------
+
+.. attribute:: Index.condition
+
+.. versionadded:: 2.2
+
+The index can be customized to include a condition. If the table is very large,
+and your queries mostly target a subset, it's convenient to restrict the index
+to that subset.
+
+:attr:`Index.name` is required when using ``condition``.
+
+.. admonition:: Restrictions on PostgreSQL
+
+    PostgreSQL requires functions referenced in the condition to be be marked as
+    IMMUTABLE. Django does not perform any checks on this, and an error will be
```

Comment:
Would it be possible to mark functions in Django so we can perform checks for this?

Reply:
The same issue with mutable functions is also present when you make a functional index. PostgreSQL's system tables contain a list of the available functions as well as a flag that indicates if it's immutable. https://dba.stackexchange.com/questions/144839/concat-used-in-index-causes-error-functions-in-index-expression-must-be-marked#144937 has the SQL for retrieving parts of the `pg_proc` catalog, which is quite extensive.